src/js/data/ui-update/index.ts

```diff
@@ -0,0 +1,131 @@
+import { Observable, of, throwError } from "rxjs";
+import {
+  exhaustMap,
+  map,
+  publishReplay,
+  refCount,
+  retry
+} from "rxjs/operators";
+import { RequestResponse } from "@dcos/http-service";
+import { makeExecutableSchema } from "graphql-tools";
+import { DCOSUIUpdateClient, UIVersionResponse } from "dcos-ui-update-client";
+
+import Config from "#SRC/js/config/Config";
+import { UIVersionSchema } from "#SRC/js/data/ui-update/UIVersion";
+
+export interface ResolverArgs {
+  uiUpdateClient: DCOSUIUpdateClient;
+}
+
+export interface GeneralArgs {
+  [key: string]: any;
+}
+
+export interface QueryContext {
+  [key: string]: any;
+}
+
+export interface UIUpdateArgs {
+  newVersion: string;
+}
+
+function isUIUpdateArgs(args: GeneralArgs): args is UIUpdateArgs {
+  return (args as UIUpdateArgs).newVersion !== undefined;
+}
+
+export function resolvers({ uiUpdateClient }: ResolverArgs) {
+  return {
+    UIVersion: {
+      clientBuild() {
+        // @ts-ignore
+        return of(window.DCOS_UI_VERSION);
+      },
+      packageVersion(
+        _parent = {},
+        _args = {},
+        context: { _versionStream: Observable<any> }
+      ) {
+        const version$ = context._versionStream;
+        return version$.pipe(
+          map(
+            ({ response }: RequestResponse<UIVersionResponse>) =>
+              response.packageVersion
+          )
+        );
+      },
+      packageVersionIsDefault(
+        _parent = {},
+        _args = {},
+        context: { _versionStream: Observable<any> }
+      ) {
+        const version$ = context._versionStream;
+        return version$.pipe(
+          map(
+            ({ response }: RequestResponse<UIVersionResponse>) =>
+              response.default
+          )
+        );
+      }
+    },
+    Query: {
+      ui(_parent = {}, _args = {}, context: QueryContext = {}) {
+        context._versionStream = of({}).pipe(
+          exhaustMap(uiUpdateClient.fetchVersion),
+          retry(2),
+          publishReplay(1),
+          refCount()
+        );
+        return of({});
+      }
+    },
+    Mutation: {
+      updateDCOSUI(_parent = {}, args: GeneralArgs = {}) {
+        if (!isUIUpdateArgs(args)) {
+          return throwError(
+            "updateDCOSUI arguments aren't valid for type UIUpdateArgs"
+          );
+        }
+
+        return uiUpdateClient
+          .updateUIVersion(args.newVersion)
+          .pipe(
+            map(
+              ({ response }: RequestResponse<string>) => `Complete: ${response}`
+            )
+          );
+      },
+      resetDCOSUI() {
+        return uiUpdateClient
+          .resetUIVersion()
+          .pipe(
+            map(
+              ({ response }: RequestResponse<string>) => `Complete: ${response}`
+            )
+          );
+      }
+    }
+  };
+}
+
+const baseSchema = `
+type Query {
+  ui: UIVersion!
```

Comment:
Nitpick, but isn't it more of a set of metadata for the ui that we are modelling here?

Reply:
I like `UIMetadata` as a name here much more. 👍 thank you, naming things is hard.

Reply:
👍 awesome :)